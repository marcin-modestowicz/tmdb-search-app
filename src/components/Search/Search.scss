.root {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;

  .typeahead {
    margin-right: 1em;
    margin-left: 1em;
    width: 12em;
    margin-top: 1em;
  }

  .options {
    margin-top: 1em;
    display: flex;
  }

  .dropdown {
    margin-right: 1em;
    width: 6em;
  }

  .button {
    width: 5em;
  }
}
