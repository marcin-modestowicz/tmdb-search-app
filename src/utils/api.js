// @flow
import request from "./request";

class Api {
  search(
    query: string,
    type: "multi" | "movie" | "person" | "tv"
  ): Promise<Object> {
    return request
      .get({
        url: `/search/${type}`,
        data: {
          query
        }
      })
      .then(response => response.json());
  }
}

export default new Api();
