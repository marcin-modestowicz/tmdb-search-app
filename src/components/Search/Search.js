// @flow

import React, { Component } from "react";
import { observable, action } from "mobx";
import { observer } from "mobx-react";
import api from "../../utils/api";
import type { FieldValue } from "../../types";
import Typeahead from "../Typeahead";
import Dropdown from "../Dropdown";
import Button from "../Button";
import styles from "./Search.css";

type Props = {
  onSearch: (Object[]) => void
};

@observer
class Search extends Component<Props> {
  searchScopeOptions: FieldValue = [
    {
      value: "multi",
      label: "All"
    },
    {
      value: "movie",
      label: "Movie"
    },
    {
      value: "tv",
      label: "TV Show"
    },
    {
      value: "person",
      label: "People"
    }
  ];
  @observable searchValue: string = "";
  @observable
  searchScope: FieldValue = {
    value: "multi",
    label: "All"
  };

  @action
  handleSuggestionChange = (value: string) => {
    this.searchValue = value;
  };

  @action
  handleSuggestionSelect = (suggestion: FieldValue) => {
    this.searchValue = suggestion.label;
    this.props.onSearch([suggestion]);
  };

  @action
  getSuggestions = (query: string) => {
    return api
      .search(query, this.searchScope.value)
      .then((response: { results: Object[] }) =>
        response.results.map(result => {
          const name = result.name || result.original_title;

          return Object.assign(result, { value: name, label: name });
        })
      );
  };

  @action
  handleSearchScopeChange = (value: string) => {
    this.searchScope = this.searchScopeOptions.find(
      option => option.value === value
    );
  };

  @action
  handleButtonClick = async () => {
    const result = await api.search(this.searchValue, this.searchScope.value);

    this.props.onSearch(result.results);
  };

  render() {
    return (
      <div className={styles.root}>
        <div className={styles.typeahead}>
          <Typeahead
            value={this.searchValue}
            onChange={this.handleSuggestionChange}
            onSelect={this.handleSuggestionSelect}
            dataSource={this.getSuggestions}
            placeholder="Search"
          />
        </div>
        <div className={styles.options}>
          <div className={styles.dropdown}>
            <Dropdown
              value={this.searchScope.label}
              options={this.searchScopeOptions}
              onSelect={this.handleSearchScopeChange}
            />
          </div>
          <div className={styles.button}>
            <Button onClick={this.handleButtonClick}>Search</Button>
          </div>
        </div>
      </div>
    );
  }
}

export default Search;
