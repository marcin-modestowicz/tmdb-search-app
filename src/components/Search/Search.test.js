// @flow
import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import api from "../../utils/api";
import Search from "./Search";

describe("Search component", () => {
  let apiMock;
  let response;
  let search;
  let onSearch;
  let instance;

  beforeAll(() => {
    apiMock = jest.spyOn(api, "search");
    response = {
      results: [{ id: "1", original_title: "test" }, { id: "2", name: "test2" }]
    };
    apiMock.mockImplementation(() => Promise.resolve(response));
  });

  afterAll(() => {
    apiMock.mockRestore();
  });

  beforeEach(() => {
    onSearch = jest.fn();
    search = shallow(<Search onSearch={onSearch} />);
    instance = search.instance();
  });

  test("should render properly", () => {
    expect(toJson(search)).toMatchSnapshot();
  });

  test("method handleSuggestionChange should update searchValue", () => {
    instance.handleSuggestionChange("test");

    expect(instance.searchValue).toBe("test");
  });

  describe("method handleSuggestionSelect", () => {
    let suggestion;

    beforeEach(() => {
      suggestion = { value: "test", label: "Test" };

      instance.handleSuggestionSelect(suggestion);
    });

    test("should update searchValue", () => {
      expect(instance.searchValue).toBe("Test");
    });

    test("should call onChange prop", () => {
      expect(onSearch).toHaveBeenCalledWith(suggestion);
    });
  });

  describe("method getSuggestions", () => {
    let result;

    beforeEach(async () => {
      result = await instance.getSuggestions("query", "multi");
    });

    test("should query api and transform suggestions", () => {
      expect(apiMock).toHaveBeenCalledWith("query", "multi");
    });

    test("should transform suggestions", () => {
      expect(result).toEqual([
        { id: "1", original_title: "test", value: "test", label: "test" },
        { id: "2", name: "test2", value: "test2", label: "test2" }
      ]);
    });
  });

  test("method handleSearchScopeChange should update searchScope", () => {
    instance.handleSearchScopeChange("person");

    expect(instance.searchScope).toEqual({
      value: "person",
      label: "People"
    });
  });

  describe("method handleButtonClick", async () => {
    beforeEach(async () => {
      instance.searchValue = "John Doe";
      instance.searchScope = {
        value: "person",
        label: "People"
      };

      await instance.handleButtonClick();
    });

    test("should query api", () => {
      expect(apiMock).toHaveBeenCalledWith("John Doe", "person");
    });

    test("should call onChange prop", () => {
      expect(onSearch).toHaveBeenCalledWith(response);
    });
  });
});
