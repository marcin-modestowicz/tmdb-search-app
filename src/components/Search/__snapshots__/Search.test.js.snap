// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component should render properly 1`] = `
<div
  className="root"
>
  <div
    className="typeahead"
  >
    <Typeahead
      dataSource={[Function]}
      onChange={[Function]}
      onSelect={[Function]}
      placeholder="Search"
      value=""
    />
  </div>
  <div
    className="options"
  >
    <div
      className="dropdown"
    >
      <Dropdown
        onSelect={[Function]}
        options={
          Array [
            Object {
              "label": "All",
              "value": "multi",
            },
            Object {
              "label": "Movie",
              "value": "movie",
            },
            Object {
              "label": "TV Show",
              "value": "tv",
            },
            Object {
              "label": "People",
              "value": "person",
            },
          ]
        }
        value="All"
      />
    </div>
    <div
      className="button"
    >
      <Button
        onClick={[Function]}
      >
        Search
      </Button>
    </div>
  </div>
</div>
`;
